micrometer-tracing/src/main/java/io/micrometer/tracing/contextpropagation/BaggageThreadLocalAccessor.java

```diff
@@ -0,0 +1,237 @@
+/*
+ * Copyright 2024 VMware, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micrometer.tracing.contextpropagation;
+
+import io.micrometer.common.lang.NonNull;
+import io.micrometer.common.util.internal.logging.InternalLogger;
+import io.micrometer.common.util.internal.logging.InternalLoggerFactory;
+import io.micrometer.context.ThreadLocalAccessor;
+import io.micrometer.tracing.BaggageInScope;
+import io.micrometer.tracing.Span;
+import io.micrometer.tracing.Tracer;
+import io.micrometer.tracing.Tracer.SpanInScope;
+
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Consumer;
+
+/**
+ * {@link ThreadLocalAccessor} used to propagate baggage via {@link BaggageToPropagate}.
+ *
+ * @since 1.2.2
+ * @author Marcin Grzejszczak
+ */
+public class BaggageThreadLocalAccessor implements ThreadLocalAccessor<BaggageToPropagate> {
+
+    private static final InternalLogger log = InternalLoggerFactory.getInstance(BaggageThreadLocalAccessor.class);
+
+    final Map<Thread, BaggageAndScope> baggageInScope = new ConcurrentHashMap<>();
+
+    /**
+     * Key under which Micrometer Tracing Baggage accessor is being registered.
+     */
+    public static final String KEY = "micrometer.tracing.baggage";
+
+    private final Tracer tracer;
+
+    /**
+     * Creates a new instance of this class.
+     * @param tracer tracer
+     * @since 1.2.2
+     */
+    public BaggageThreadLocalAccessor(Tracer tracer) {
+        this.tracer = tracer;
+    }
+
+    @Override
+    public Object key() {
+        return KEY;
+    }
+
+    @Override
+    public BaggageToPropagate getValue() {
+        Map<String, String> baggage = tracer.getAllBaggage();
+        if (log.isTraceEnabled()) {
+            log.trace("Current baggage in thread local [" + baggageInScope.get(Thread.currentThread())
+                    + "], current baggage from tracer [" + baggage + "]");
+        }
+        return (baggage == null || baggage.isEmpty()) ? null : new BaggageToPropagate(baggage);
```

Comment:
I'm a bit concerned about the `isEmpty()` check. I can imagine a situation where even though the map is empty, it can be filled in later, which won't be reflected. That sounds ok, but at the same time, currently, a non-empty map can become empty later and with a reference used will make the contents be reflected in the captured `BaggageToPropagate`. This should be consistent. Most probably by just copying the contents in the constructor of the holder.